src/hazelcore/dynamics/Environment.re

```diff
@@ -1,3 +1,136 @@
+  | InvalidText(_) => d2
+  | Keyword(_) => d2
+  | Let(dp, d3, d4) =>
+    let d3 = subst_var(d1, x, d3);
+    let d4 =
+      if (DHPat.binds_var(x, dp)) {
+        d4;
+      } else {
+        subst_var(d1, x, d4);
+      };
+    Let(dp, d3, d4);
+  | FixF(y, ty, d3) =>
+    let d3 =
+      if (Var.eq(x, y)) {
+        d3;
+      } else {
+        subst_var(d1, x, d3);
+      };
+    FixF(y, ty, d3);
+  | Lam(dp, ty, d3) =>
+    if (DHPat.binds_var(x, dp)) {
+      d2;
+    } else {
+      let d3 = subst_var(d1, x, d3);
+      Lam(dp, ty, d3);
+    }
+  | Ap(d3, d4) =>
+    let d3 = subst_var(d1, x, d3);
+    let d4 = subst_var(d1, x, d4);
+    Ap(d3, d4);
+  | BoolLit(_)
+  | IntLit(_)
+  | FloatLit(_)
+  | ListNil(_)
+  | Triv => d2
+  | Cons(d3, d4) =>
+    let d3 = subst_var(d1, x, d3);
+    let d4 = subst_var(d1, x, d4);
+    Cons(d3, d4);
+  | BinBoolOp(op, d3, d4) =>
+    let d3 = subst_var(d1, x, d3);
+    let d4 = subst_var(d1, x, d4);
+    BinBoolOp(op, d3, d4);
+  | BinIntOp(op, d3, d4) =>
+    let d3 = subst_var(d1, x, d3);
+    let d4 = subst_var(d1, x, d4);
+    BinIntOp(op, d3, d4);
+  | BinFloatOp(op, d3, d4) =>
+    let d3 = subst_var(d1, x, d3);
+    let d4 = subst_var(d1, x, d4);
+    BinFloatOp(op, d3, d4);
+  | Inj(ty, side, d3) =>
+    let d3 = subst_var(d1, x, d3);
+    Inj(ty, side, d3);
+  | Pair(d3, d4) =>
+    let d3 = subst_var(d1, x, d3);
+    let d4 = subst_var(d1, x, d4);
+    Pair(d3, d4);
+  | ConsistentCase(Case(d3, rules, n)) =>
+    let d3 = subst_var(d1, x, d3);
+    let rules = subst_var_rules(d1, x, rules);
+    ConsistentCase(Case(d3, rules, n));
+  | InconsistentBranches(u, i, sigma, Case(d3, rules, n)) =>
+    let d3 = subst_var(d1, x, d3);
+    let rules = subst_var_rules(d1, x, rules);
+    let sigma' = subst_var_env(d1, x, sigma);
+    InconsistentBranches(u, i, sigma', Case(d3, rules, n));
+  | EmptyHole(u, i, sigma) =>
+    let sigma' = subst_var_env(d1, x, sigma);
+    EmptyHole(u, i, sigma');
+  | NonEmptyHole(reason, u, i, sigma, d3) =>
+    let d3' = subst_var(d1, x, d3);
+    let sigma' = subst_var_env(d1, x, sigma);
+    NonEmptyHole(reason, u, i, sigma', d3');
+  | Cast(d, ty1, ty2) =>
+    let d' = subst_var(d1, x, d);
+    Cast(d', ty1, ty2);
+  | FailedCast(d, ty1, ty2) =>
+    let d' = subst_var(d1, x, d);
+    FailedCast(d', ty1, ty2);
+  | InvalidOperation(d, err) =>
+    let d' = subst_var(d1, x, d);
+    InvalidOperation(d', err);
+  }
+and subst_var_rules =
+    (d1: DHExp.t, x: Var.t, rules: list(DHExp.rule)): list(DHExp.rule) =>
+  rules
+  |> List.map((r: DHExp.rule) =>
+       switch (r) {
+       | Rule(dp, d2) =>
+         if (DHPat.binds_var(x, dp)) {
+           r;
+         } else {
+           Rule(dp, subst_var(d1, x, d2));
+         }
+       }
+     )
+and subst_var_env = (d1: DHExp.t, x: Var.t, sigma: t): t =>
+  sigma
+  |> List.map(xd => {
+       let (y, d) = xd;
+       (y, subst_var(d1, x, d));
+     });
+
+let subst = (env: t, d: DHExp.t): DHExp.t =>
```

Comment:
substitution stuff should be in Evaluator or in a separate Subst module, not part of Environment